### PR TITLE
Add TCD scheduler

### PR DESCRIPTION
**Pull request by l0stl0rd**
Adds the TCD scheduler to better support.
https://huggingface.co/h1t/TCD-SDXL-LoRA or checkpoints that have been made with TCD

Example:
TCD Lora with Euler A
b0ad6174-cd2b-49fe-ae42-3a83bc6ae571

TCD Lora with TCD scheduler
74495a51-eeac-45e6-9983-fb6551a5bdef

### CODE REVIEW

**Comment by psychedelicious** [comment]
Thanks for the PR. I've tidied a couple things.

The scheduler works fine, but I'm not sure we fully support TCD at this time. I noticed a couple things when reviewing the TCD repo:
- The examples use a guidance of 0. This is unexpected. The app currently doesn't allow a value under 1. We can certainly change that if there's a reason to do so, but it's not clear if that exists.
- The examples add an `eta` arg to the diffusers pipeline and makes it sound like this should be user-configurable. It's not clear if this is really required or not.

I'm able to use the TCD SDXL LoRA, but it doesn't produce good results with the scheduler. I'm not sure if this is because I can only go as low as 1 for guidance, or the `eta` arg isn't provided, or if there is some other factor related to our diffusers implementation for the scheduler or LoRA.

Example output with the LoRA, 8 steps, CFG 1, per the examples:
image

(the higher the CFG, the worse the results)

Example output without the LoRA, 

The scheduler _does_ work fine without the TCD LoRA and with "normal" settings:
image

I'm reluctant to add the scheduler without the other changes to support the TCD model, because it communicates that we support hte TCD model (which appears to not untrue).

**Comment by l0stl0rd** [comment]
Thanks, it was the first time I tried adding something.

I noticed the same thing with the CFG for example, that is why I said on Discord, “it works but bot sure if it is correct.

I noticed the eta thing but was not sure what to do with it.

**Comment by psychedelicious** [comment]
@l0stl0rd Sorry, I think I missed that discord thread. Can you link me to it? Can discuss the rest of the implementation there.

And thanks for the link to vlad's implementation, that will be a useful reference.

**Comment by l0stl0rd** [comment]
it was not in the dev chat but here it is https://discord.com/channels/1020123559063990373/1020123559831539744/1222968707240689694

You will see that I did not get EDM to work :(

**Comment by blessedcoolant** [comment]
The implementation in this PR is fine. The only issue I noticed is that when applying the LoRA, we're still getting a pretty noisy output at the default eta of 0.3 which is what Diffusers ships with. If we manually change that to 1.0, TCD is functioning as it should.

Lower values are okay too but from the little experimentation I did, it seems like 1.0 is what we set the gamma too for the step calculation.

I've updated this PR with that change. Please test and let me know. If 1.0 works fine in all cases and produces expected results, then we can keep it at that. Otherwise we'll have to expose this value for just that scheduler.

**Review by blessedcoolant — APPROVED** [review]

**Comment by blessedcoolant** [comment]
This PR is fine to merge I think. Even the Hyper SDXL paper that utilizes the TCD Scheduler prefers the gamma value to be at 1.0 and as per my testing it holds up.

They released a new model today that allows guidance up to 5-7 with their new LoRA. That kinda fails for some reason but that is not related to this PR. I'll take a look at why that LoRA is not working correctly. The rest of the HyperSD LoRA's work just fine with the new TCD Scheduler.

File: invokeai/app/invocations/latent.py [hunk 1]
```diff
@@ -853,13 +855,15 @@ def init_scheduler(
...
+        scheduler_step_kwargs: Dict[str, Any] = {}
         scheduler_step_signature = inspect.signature(scheduler.step)
         if "generator" in scheduler_step_signature.parameters:
             # At some point, someone decided that schedulers that accept a generator should use the original seed with
             # all bits flipped. I don't know the original rationale for this, but now we must keep it like this for
             # reproducibility.
-            scheduler_step_kwargs = {"generator": torch.Generator(device=device).manual_seed(seed ^ 0xFFFFFFFF)}
+            scheduler_step_kwargs.update({"generator": torch.Generator(device=device).manual_seed(seed ^ 0xFFFFFFFF)})
+        if "eta" in scheduler_step_signature.parameters and isinstance(scheduler, TCDScheduler):
+            scheduler_step_kwargs.update({"eta": 1.0})
```
Comment (psychedelicious) [inline invokeai/app/invocations/latent.py:866]: Don't we only need to check that it's an instance of `TCDScheduler` (and not that `eta` is in the parameters)?

Comment (blessedcoolant) [reply invokeai/app/invocations/latent.py:866]: Umm yeah.. but it doesn't hurt for the double check I guess. Can remove. Not an issue.

**Review by psychedelicious — APPROVED** [review]
Sorry was just curious about the `eta` check, didn't mean to block this

**Comment by blessedcoolant** [comment]
> Sorry was just curious about the `eta` check, didn't mean to block this

Na you're right. Initially I did the eta check coz I thought it was only on the TCD but then I realized there were other schedulers that used gamma variance too. So added in a specific check for TCD. And given it is a keyword argument, we don't really need to check if it exists or not. If it ever gets removed upstream for whatever reason, given it is a kwarg, it shouldn't break the app on our end. But having the double check might. So updated that.
